File: main1.c
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <string.h>

static char *
foo(uint64_t user_id, const char *login)
{
    char *x = NULL, *suffix = NULL;
    asprintf(&suffix, " [%s]", ({
        user_id
            ? ({ asprintf(&x, "user_id = %llu", user_id); })
            : ({ asprintf(&x, "login = \"%s\"", login); });
        x;
    }));

    free(x);

    char *r = NULL;
    if (strlen(login) > 10)
    {
        asprintf(&r, "bad login %s", suffix);

        free(suffix);
        return r;
    }

    printf("action is complete %s", suffix);

    free(suffix);
    return strdup("");
}

int main()
{
    char *r = foo(10, "VALID");
    free(r);
}

File: main1.rs
fn foo(user_id: u64, login: &str) -> String {
    let suffix = format!(
        " [{}]",
        (|| {
            if user_id != 0 {
                format!("user_id = {}", user_id)
            } else {
                format!(r#"login = "{}""#, login)
            }
        })()
    );

    if login.len() > 10 {
        return format!("bad login: {}", suffix);
    }

    println!("action is complete {}", suffix);
    String::from("")
}

fn main() {
    foo(10, "VALID");
}
